src/view/treeNodes/directoryTreeNode.ts

```diff
@@ -109,22 +110,40 @@ export class DirectoryTreeNode extends TreeNode implements vscode.TreeItem2 {
 
 		let node = this.pathToChild.get(dir);
 		if (!node) {
-			node = new DirectoryTreeNode(this, dir);
+			node = new DirectoryTreeNode(this, dir, this.pullRequestModel);
 			this.pathToChild.set(dir, node);
 			this.children.push(node);
 		}
 
 		node.addPathRecc(tail, file);
 	}
 
-	updateCheckbox(newState: vscode.TreeItemCheckboxState) {
-		this.children.forEach(child => child.updateCheckbox(newState));
+	updateCheckbox(newState: vscode.TreeItemCheckboxState, performMutation: boolean) {
+		if (performMutation) {
+			const filenames = this.gatherPaths();
+
+			if (newState === vscode.TreeItemCheckboxState.Checked) {
+				this.pullRequestModel.markFilesAsViewed(filenames);
+			} else {
+				this.pullRequestModel.unmarkFilesAsViewed(filenames);
+			}
+		}
+
+		this.children.forEach(child => child.updateCheckbox(newState, false));
 
 		if (this.parent instanceof TreeNode && !this.parent.updateParentCheckbox()) {
 			this.refresh(this);
 		}
 	}
 
+	// recursively gathers the paths of all files in this directory and the directories below
+	private gatherPaths() {
+		const subDirectories = this.children.filter((child): child is DirectoryTreeNode => child instanceof DirectoryTreeNode);
+		const paths = this.children.filter((child): child is RemoteFileChangeNode | InMemFileChangeNode | GitFileChangeNode => !(child instanceof DirectoryTreeNode)).map(child => child.fileName);
```

Comment:
Here we're now going through all children (of which there could be many) 2x, when we could instead iterate through once and add each child to `subDirectories` or `paths` as appropriate.